### ISSUE TITLE
P256 field addition, subtraction, negation

### ISSUE THREAD

**Issue by andres-erbsen** [mit-plv/fiat-crypto#218]
add https://github.com/openssl/openssl/blob/8fc063dcc9668589fd95533d25932396d60987f9/crypto/ec/asm/ecp_nistz256-x86_64.pl#L330
sub https://github.com/openssl/openssl/blob/8fc063dcc9668589fd95533d25932396d60987f9/crypto/ec/asm/ecp_nistz256-x86_64.pl#L390
opp https://github.com/openssl/openssl/blob/8fc063dcc9668589fd95533d25932396d60987f9/crypto/ec/asm/ecp_nistz256-x86_64.pl#L442

**Comment by JasonGross** [comment]
I've made some progress on this in https://github.com/mit-plv/fiat-crypto/commit/f79880fc1f986c7b46a185c90a70233fef9698fb.  I hope to have integration tests and display files (which will produce wrong code for `add`, and will be mostly admitted for `sub` and `opp`) for them later tonight (if not, then by the end of the day Monday).

@jadephilipoom The context variables I'll need you to fill are:
```coq
sub_then_maybe_add_cps_id : forall (R_numlimbs : nat) (Av Bv : Saturated.T R_numlimbs) (cpsT : Type)
                              (f : Saturated.T R_numlimbs -> cpsT), sub_then_maybe_add_cps Av Bv f = f (sub_then_maybe_add Av Bv)
sub_then_maybe_add_cps : forall R_numlimbs : nat,
                         Saturated.T R_numlimbs ->
                         Saturated.T R_numlimbs -> forall cpsT : Type, (Saturated.T R_numlimbs -> cpsT) -> cpsT
sub_then_maybe_add : forall R_numlimbs : nat, Saturated.T R_numlimbs -> Saturated.T R_numlimbs -> Saturated.T R_numlimbs
small_sub_then_maybe_add : forall (r : positive) (R_numlimbs : nat) (N a b : Saturated.T R_numlimbs),
                           Saturated.small (Z.pos r) a ->
                           Saturated.small (Z.pos r) b ->
                           (0 <= Saturated.eval (Z.pos r) a < Saturated.eval (Z.pos r) N)%Z ->
                           (0 <= Saturated.eval (Z.pos r) b < Saturated.eval (Z.pos r) N)%Z ->
                           Saturated.small (Z.pos r) (sub_then_maybe_add a b)
eval_sub_then_maybe_add : forall (r : positive) (R_numlimbs : nat) (N a b : Saturated.T R_numlimbs),
                          Saturated.small (Z.pos r) a ->
                          Saturated.small (Z.pos r) b ->
                          (0 <= Saturated.eval (Z.pos r) a < Saturated.eval (Z.pos r) N)%Z ->
                          (0 <= Saturated.eval (Z.pos r) b < Saturated.eval (Z.pos r) N)%Z ->
                          Saturated.eval (Z.pos r) (sub_then_maybe_add a b) =
                          (Saturated.eval (Z.pos r) a - Saturated.eval (Z.pos r) b +
                           (if Saturated.eval (Z.pos r) a - Saturated.eval (Z.pos r) b <? 0 then Saturated.eval (Z.pos r) N else
0))%Z
```

**Comment by JasonGross** [comment]
I've finished making (partially admitted, partially wrong) display files; #221 will remove the admits once `sub_then_maybe_add_cps` and `sub_then_maybe_add` are defined and used.

**Comment by jadephilipoom** [comment]
`sub_then_maybe_add` now exists : aea636

**Comment by jadephilipoom** [comment]
I'm not completely confident that I got all the numbers of limbs right everywhere, so we should check that it simplifies to the correct thing. I'll be free again this evening to fix it if not.

**Comment by JasonGross** [comment]
@jadephilipoom or @andres-erbsen can one of you instantiate the axioms?  If not, I'll plan to do it at around 8 or 9

**Comment by jadephilipoom** [comment]
Done

**Comment by JasonGross** [comment]
The only thing this is blocking on now is a proper ripple-carry-add.

**Comment by JasonGross** [comment]
I believe this will be closed by #226.
